[flutter_tools] Introducing arg option for specifying the output directory for web

eliasyishak opened the pull request:
Adding cli arg option that will allow developers to specify where they would like their build artifacts for web.

Addresses issue: https://github.com/flutter/flutter/issues/71130

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

eliasyishak committed:
introducing arg option for specifying the output directory

eliasyishak committed:
fixing analyze error

eliasyishak committed:
new test added for checking specified build dir for web

christopherfujino commented on packages/flutter_tools/lib/src/commands/build_web.dart:
```diff
@@ -67,6 +67,11 @@ class BuildWebCommand extends BuildSubCommand {
           'Valid values range from O0 to O4.'
     );
+    argParser.addOption(
+        'output-dir',
```
1. can we match what `dart compile exe` does, which is have the option be named `--output` with a shortcut `-o`?
2. does this have to be an absolute path? can it be relative, like, `--ouput=../build/foo`?
3. are there any limitations to supporting this for every target platform?

eliasyishak replied:
1 & 2: I was actually trying to follow the same pattern we had implemented for from `build_aar.dart` where they are using an absolute path with the same option definition. Should we instead follow what `dart compile exe` does instead?

3. I would need to do additional testing but I imagine if it has been working well in `build_aar.dart` and now web, perhaps we may be able to do it for every platform

christopherfujino replied:
oh wow, we actually implemented this for build_aar?

I would say for any platform we can trivially implement this for we should. and if there are any that take more than a few lines of code, we should track implementing with a TODO and indicate from the option help text that it is not yet implemented.

eliasyishak replied:
Following up from chat conversation; I will mark a TODO for implementing this feature across other build targets

christopherfujino commented on packages/flutter_tools/lib/src/web/compile.dart:
```diff
@@ -28,10 +28,13 @@ Future<void> buildWeb(
-  final Directory outputDirectory = globals.fs.directory(getWebBuildDirectory());
+  final Directory outputDirectory = outputDirectoryPath == null
+      ? globals.fs.directory(getWebBuildDirectory())
+      : globals.fs.directory(outputDirectoryPath);
```
Will we recursively create the directory if some of its parents don't already exist?

eliasyishak replied:
Yes we will, the `recursive` parameter has been marked as true in the line below

`outputDirectory.createSync(recursive: true);`

christopherfujino commented on packages/flutter_tools/test/commands.shard/hermetic/build_web_test.dart:
```diff
@@ -116,6 +116,46 @@ void main() {
+    setupFileSystemForEndToEndTest(fileSystem);
+
+    const String newBuildDir = 'new_dir';
+    final Directory buildDir = fileSystem.directory(fileSystem.path.join(newBuildDir));
```
can we test here that the directory does NOT exist?

christopherfujino commented on packages/flutter_tools/test/commands.shard/hermetic/build_web_test.dart:
```diff
@@ -116,6 +116,46 @@ void main() {
+      '--output-dir=$newBuildDir'
+    ]);
+
+    expect(buildDir.existsSync(), true);
```
can you also check some of the contents of the build dir?

eliasyishak replied:
I'm not sure the test setup actually produces anything in the build directory. When I use the debugger to inspect the contents of that directory, it looks like it is empty. I followed the same setup used for this test shown here

https://github.com/eliasyishak/flutter/blob/specify-build-output-directory/packages/flutter_tools/test/commands.shard/hermetic/build_web_test.dart#L87:L117

Should I continue digging into the test build system or do you think this is enough?

christopherfujino replied:
so are we just testing that the buildDir gets created? do you know why we don't put anything in it?

eliasyishak replied:
yes, just finished chatting with @jonahwilliams and he suggests that we test that the build directory is correctly getting created and configured and allow the build system and its tests to correctly identify any problems that come during the build

christopherfujino replied:
ok, can you update the test name to reflect what we're actually testing here (which seems to me to be that the buildDir is created at the correct location)?

eliasyishak committed:
add'l test added to check that dir doesnt exist before build

eliasyishak committed:
cleanup

eliasyishak committed:
correctly format todo

christopherfujino commented on packages/flutter_tools/lib/src/web/compile.dart:
```diff
@@ -28,10 +28,13 @@ Future<void> buildWeb(
   bool nativeNullAssertions,
   String? baseHref,
   String? dart2jsOptimization,
+  [String? outputDirectoryPath,]
```
Use a named parameter, because otherwise things will get awkward the next time we add an optional parameter

```suggestion
  {String? outputDirectoryPath,}
```

christopherfujino commented on packages/flutter_tools/lib/src/commands/build_web.dart:
```diff
@@ -67,6 +67,11 @@ class BuildWebCommand extends BuildSubCommand {
           'Valid values range from O0 to O4.'
     );
+    argParser.addOption(
+        'output-dir',
```
can you:

1. make this a helper method on the FlutterCommand class `usesOutputDir()` and share it with build_aar
2. make the option name `output` (to support the legacy build_aar use case, you can set `aliases: 'output-dir',`) and `abbr: 'o'` which will allow either --output=/path/to/dir or -o /path/to/dir

eliasyishak replied:
Good point, I was looking for somewhere upstream that I can add this to work for both `aar` and `web` builds. This should allow us to incrementally add for all the build subcommands

eliasyishak committed:
adding output directory upstream of build subcommands

eliasyishak committed:
clean up to help text

eliasyishak committed:
fix analyze bug

eliasyishak committed:
allow nullable output dir for build_aar

christopherfujino commented on packages/flutter_tools/lib/src/commands/build_web.dart:
```diff
@@ -113,6 +114,16 @@ class BuildWebCommand extends BuildSubCommand {
+    // Currently supporting options [output-dir] and [output]
+    // as valid approaches for setting output directory of build artifacts
+    final String? outputDirectoryPath;
+    if (stringArg('output-dir') != null) {
```
When I tried this locally, providing `flutter build web --output-dir=some-path` and then read `stringArg('output')`, I was able to see `'some-path'`

christopherfujino replied:
(note I'm querying for the real name which provides the alias result--the transitive may not be true)

eliasyishak replied:
hmm yeah looks like I must have been mistaken.. when i called `stringArg('output')`, it correctly retrieves the correct path regardless of whether i used the `output-dir` or `output` option

I also realized I was getting some weird behavior because of the two different dart versions on my computer as we discussed today which were cleared after removing the homebrew installed dart

Great catch!

eliasyishak committed:
removing conditional to check two different arg options

Caused by different dart binaries being used on my workstation that caused some strange behavior for argparsing

christopherfujino reviewed (approved):
LGTM thanks! I'll start using this immediately :)

eliasyishak committed:
clean up

eliasyishak committed:
Merge remote-tracking branch 'upstream/master' into specify-build-output-directory
